scripts/train.py

```diff
@@ -0,0 +1,181 @@
+from torch.optim import Adam
+from torch import nn, optim
+import torch
+import math
+import time
+import os
+import sys
+
+
+'''
+Decouple the reference path.
+'''
+cwd = os.getcwd()  # current working directory
+cfp = os.path.dirname(os.path.abspath(__file__))  # current file path
+os.chdir(cfp)
+sys.path.append(os.path.abspath(".."))
+os.chdir(cwd)
+
+from nn_modules.transformer.top_former import TopFormer
+from utils.config import top_config
+from data.data_loader import train_loader, val_loader
+
+model = TopFormer(num_layers=top_config["model_params"]["num_layers"], d_model=top_config["model_params"]
+                  ["d_model"], nhead=top_config["model_params"]["nhead"], num_classes=top_config["model_params"]["num_classes"])
+
+# TODO(diwei): remove
+
+
+def count_parameters(model):
+    return sum(p.numel() for p in model.parameters() if p.requires_grad)
+
+
+# TODO(diwei): remove
+def initialize_weights(m):
+    if hasattr(m, 'weight') and m.weight.dim() > 1:
+        nn.init.kaiming_uniform(m.weight.data)
+
+
+# TODO(diwei): remove
+def epoch_time(start_time, end_time):
+    elapsed_time = end_time - start_time
+    elapsed_mins = int(elapsed_time / 60)
+    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
+    return elapsed_mins, elapsed_secs
+
+
+# TODO(diwei): remove
+print(f'The model has {count_parameters(model):,} trainable parameters')
+model.apply(initialize_weights)
+
+
+# TODO(diwei): adjust
+optimizer = Adam(params=model.parameters(),
+                 lr=top_config["opt_params"]["lr"],
+                 weight_decay=top_config["opt_params"]["weight_decay"],
+                 eps=top_config["opt_params"]["eps"])
+
+scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer=optimizer,
+                                                 verbose=True)
+
+# TODO(diwei): adjust
+criterion = nn.CrossEntropyLoss()
+
+
+def train(model, loader, optimizer, criterion, device='cpu'):
+    """THe interface to train the model using training data
+
+    Parameters
+    ----------
+    model : nn.Module
+        The TopFormer model used to train.
+    loader : DataLoader
+        The loader to visit the data.
+    optimizer : Adam
+        Optimizer to iterate the params.
+    criterion : Loss nn.CrossEntropyLoss
+        Loss criterion.
+    device : str, optional
+        Device used to evaluate, by default 'cpu'
+
+    Returns
+    -------
+    float
+        Average loss.
+    """
+    epoch_loss = 0
+    model.to(device)
+    for i, batch in enumerate(loader):
+        x = batch["x"]
+        u = batch["u"]
+        gt = batch["gt"]
+
+        optimizer.zero_grad()
+        output = model(x.permute(1, 0, 2), u.permute(1, 0, 2))
```

Comment:
The nn transformer encoder/decoder layer expect inputs in the format (seq, batch, feature) by default, which aligns with the TopDataset structure. There is no need to permute.

Reply:
The dataloader concatenates the item alone with first axis, we still permute the data with (1,0,2).

Reply:
You are right. Let's keep the permutation.